sql: implement ADD/DROP schema changes for row-level TTL

otan opened the pull request:
These commits adds the "scarier" part of the row level TTL job, in particular:
* DROP TABLE/DATABASE/SCHEMA (this required adding it to the declarative schema changer as well, see the last two commits)
* SET (ttl) for the first time, which adds the automatic column
* RESET (ttl), which drops the automatic column

This is handled in both the old and new schema changer.

otan committed:
sql: implement functionality for modifying ttl_expire_after

This commit implements changing the DEFAULT and ON UPDATE expressions
automatically whenever ttl_expire_after is changed.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76216)
<!-- Reviewable:end -->

otan committed:
sql: drop schedules on a TTL table when it is dropped

This commit drops all schedules tied to the table's TTL job when the
table is dropped through DROP TABLE/SCHEMA/DATABASE.

Release note: None

otan committed:
sql: handle adding TTL to a table during an ALTER statement

This commit handles adding a TTL to a table using
`ALTER TABLE ... SET ...`.

To accomplish this, we needed to make add column finalization and adding
the TTL struct / scheduled job atomic. We add a new `ModifyRowLevelTTL`
mutation which the same mutation ID, with approprate rollback handling
if, e.g. ADD COLUMN succeeds but setting the row-level TTL fails.

Release note: None

otan commented:
my reading of the schema change code is that I "only" need to support DROP TABLE/DROP DATABASE/DROP SCHEMA.
the way add/drop column is being used here is via ALTER TABLE ... SET/RESET, which means it would use the original schema changer anyway. is that right?

postamar reviewed (changes requested):
Thanks for putting this up for review so quickly. I think it's a good, solid change overall, as far as I can tell (I don't know much about scheduled jobs). The drop column is dodgy of course (not your doing, it always was dodgy) and that's exactly the kind of thing the declarative schema changer will be able to help with.

I do have a couple of high level comments and questions:

1. This is missing table descriptor validation checks for the TTL descriptor. I don't think they can wait until later. Every time you add a field in the TTL proto, you should have a corresponding validation check and unit tests in `tabledesc`. This will strengthen _all_ of your tests so, so much.

2. Wouldn't it be possible to do without the new mutation types? It feels like the TTL descriptor could simply have some kind of status enum indicating whether the schedule:
a. needs to be updated,
b. needs to be deleted,
c. no-op.

ccing @ajwerner for his :eyes: on this last one, he might want to weigh in on whether this is a good idea or not.

> Reviewed 3 of 3 files at r1, 14 of 14 files at r2, 9 of 9 files at r3, 6 of 6 files at r4, 5 of 5 files at r5, 2 of 2 files at r6, 3 of 3 files at r7, 3 of 3 files at r8, 11 of 11 files at r9, 2 of 2 files at r10, 5 of 5 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-:-MvPzWF2B8FdJN8EzZHY:b8vs09d)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/sql/alter_table.go, line 1765 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-MvPnprk4M6oMH3Yo4T2:-MvPnprk4M6oMH3Yo4T3:b684nj5) ([raw file](https://github.com/cockroachdb/cockroach/blob/dab825c2dc731872e85e15555d651b1a857090b1/pkg/sql/alter_table.go#L1765)):*
> ```Go
> 		// Do not have to do anything here.
> 	case before != nil && after != nil:
> 		if before.DeletionCron != after.DeletionCron {
> ```

What if before you have "" and after you have "@hourly"? Do we care about those kinds of no-ops?
___
*[pkg/sql/create_table.go, line 2342 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-MvPmfFz5uzAYPk24O1P:-MvPm4hX-V4iSdy7_3c1:b-nx86b0) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e12c83ef38219f953fd44ce05f67b145337977b/pkg/sql/create_table.go#L2342)):*
> ```Go
> // defaultTTLScheduleCron is the default cron duration for row-level TTL.
> // defaultTTLScheduleCron cannot be a cluster setting as this would involve
> // changing all existing schedules to match the new setting.
> ```

Would it make sense to enforce the `deletion_cron` field in the descriptor to be non-empty instead? I won't insist on it, but it feels like that would be more robust.
___
*[pkg/sql/schema_changer.go, line 388 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-MvPrXet-Ut11D0v2ZSZ:-MvPrXet-Ut11D0v2ZS_:b-jmnien) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c43ecc15eeb952e0f11efbcb6bb39bccc5def21/pkg/sql/schema_changer.go#L388)):*
> ```Go
> 	if tableDesc.Dropped() && tableDesc.GetRowLevelTTL() != nil {
> 		if err := sc.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
> 			scheduleID := tableDesc.GetRowLevelTTL().ScheduleID
> ```

Is this guaranteed to be non-zero at this point?
___
*[pkg/sql/schema_changer.go, line 1222 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-MvPyVkx0fNCZUAStEa7:-MvPyVkx0fNCZUAStEa8:bsts7l0) ([raw file](https://github.com/cockroachdb/cockroach/blob/cc22eac59dd8feb9f613532cb29308784d98e795/pkg/sql/schema_changer.go#L1222)):*
> ```Go
> 						}
> 						scTable.RowLevelTTL.ScheduleID = j.ScheduleID()
> 					}
> ```

nit: shouldn't this have been in an earlier commit?
___
*[pkg/sql/show_create.go, line 195 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-MvPnEw-0XQ_ETp5-1Ar:-MvPnEw-0XQ_ETp5-1As:b-av32nd) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e12c83ef38219f953fd44ce05f67b145337977b/pkg/sql/show_create.go#L195)):*
> ```Go
> 		}
> 		if cron := ttl.DeletionCron; cron != "" {
> 			storageParams = append(storageParams, fmt.Sprintf(`ttl_delete_batch_size = '%s'`, cron))
> ```

Wrong format string. I expect this will show up in tests also.
___
*[pkg/sql/catalog/descpb/structured.proto, line 612 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-MvPvql_0y98KOhZZXwm:-MvPvql_0y98KOhZZXwn:b6k1np3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ebb3f222fa9750b7a9724a44a7d4263e1cec7aa/pkg/sql/catalog/descpb/structured.proto#L612)):*
> ```protobuf
> 
>   optional TableDescriptor.RowLevelTTL row_level_ttl = 1 [(gogoproto.customname) = "RowLevelTTL"];
> }
> ```

Is this message likely to have more fields added to it? It doesn't seem necessary, to be honest.
___
*[pkg/sql/schema_changer_test.go, line 7516 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-MvPwdID5_GD2c5RKwKZ:-MvPwdID5_GD2c5RKwK_:bvliif3) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ebb3f222fa9750b7a9724a44a7d4263e1cec7aa/pkg/sql/schema_changer_test.go#L7516)):*
> ```Go
> 			setup: `
> 			CREATE DATABASE t;
> 			USE t;
> ```

FYI, these `USE t` don't seem to be necessary. I don't care much either way.


<!-- Sent from Reviewable.io -->

otan commented:
> Wouldn't it be possible to do without the new mutation types? It feels like the TTL descriptor could simply have some kind of status enum indicating whether the schedule:
> a. needs to be updated,
> b. needs to be deleted,
> c. no-op.

i can see this getting complicated because if something fails in between the reverse mutations will also need to dive deeper into the tabledescriptor proto and zero things out, but only on add column / drop column if it matches a certain column name.

for me, the separate mutation feels cleaner, but yeah more input would be good. i'll hold off on making that change (as it is not trivial :) ) for now

otan committed:
sql: make RESET (ttl) drop the TTL column

This commit makes `ALTER TABLE ... RESET (ttl)` drop the automatic TTL
column. To do this, the `DROP COLUMN` logic is extracted into a separate
function, then used along with a `ModifyRowLevelTTL` mutation.

An extra check is added to the schema changer to ensure we aren't
re-adding a schedule that doesn't yet exist.

Release note: None

postamar commented:
The way I understand it, mutations represent a future change to the table descriptor itself. It seemed to me that the TTL descriptor was effectively the source of truth for something not in the table descriptor, so to me it didn't fit into that mold. Of course, if there's subtle interplay with adding/removing the ttl column, that would invalidate my argument. In any case having a new type of mutation isn't the worst thing in the world but if it's avoidable it's better.

otan reviewed (commented):
> This is missing table descriptor validation checks for the TTL descriptor

done. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-:-MvQu-8t3lx08_dbvh9l:bexl1n9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hourly and @postamar)

___
*[pkg/sql/alter_table.go, line 1765 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-MvPnprk4M6oMH3Yo4T2:-MvQjRSU3pM6t7jtyC9V:b-52myk2) ([raw file](https://github.com/cockroachdb/cockroach/blob/dab825c2dc731872e85e15555d651b1a857090b1/pkg/sql/alter_table.go#L1765)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

What if before you have "" and after you have "@hourly"? Do we care about those kinds of no-ops?
</blockquote></details>

the net effect doesn't change much, rather not keep it too complicated for now.
___
*[pkg/sql/create_table.go, line 2342 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-MvPmfFz5uzAYPk24O1P:-MvQjVs85Gr7xW3Q9YUE:bu5xojf) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e12c83ef38219f953fd44ce05f67b145337977b/pkg/sql/create_table.go#L2342)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Would it make sense to enforce the `deletion_cron` field in the descriptor to be non-empty instead? I won't insist on it, but it feels like that would be more robust.
</blockquote></details>

i can see it both ways. i preferred this way as this paves the way for a "default" in future, and it makes the logic for displaying the cron on SHOW CREATE TABLE less gnarly if the user is using the default.
___
*[pkg/sql/schema_changer.go, line 388 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-MvPrXet-Ut11D0v2ZSZ:-MvQwv97Bct7VHk8069l:b-1u33et) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c43ecc15eeb952e0f11efbcb6bb39bccc5def21/pkg/sql/schema_changer.go#L388)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is this guaranteed to be non-zero at this point?
</blockquote></details>

In theory no, but no harm in adding a check in.
Worth noting that the deleteSchedule no-ops even if it was 0 anyway
___
*[pkg/sql/show_create.go, line 195 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-MvPnEw-0XQ_ETp5-1Ar:-MvQjfWk0ns7u50uKfKk:b-3ecymt) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e12c83ef38219f953fd44ce05f67b145337977b/pkg/sql/show_create.go#L195)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Wrong format string. I expect this will show up in tests also.
</blockquote></details>

nice catch, fixed.
___
*[pkg/sql/catalog/descpb/structured.proto, line 612 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/76216#-MvPvql_0y98KOhZZXwm:-MvQwdX97qPcq4gCErKo:bibjoyg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ebb3f222fa9750b7a9724a44a7d4263e1cec7aa/pkg/sql/catalog/descpb/structured.proto#L612)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

Is this message likely to have more fields added to it? It doesn't seem necessary, to be honest.
</blockquote></details>

how strongly do you feel about this? i guess, ideally, i would've added this into the `ColumnDescriptor` mutation with an extra field symbolizing whether to add or drop a TTL. it's easy to extend this way, and in practice i'm not sure there's too big of a difference.


<!-- Sent from Reviewable.io -->

otan committed:
catpb: move RowLevelTTL to catpb

Release note: None

otan commented:
i've tacked on two commits at the end, which adds the dropping of schedules on DROP TABLE on the new schema changer. it works in the logic test, but my go generate output generated a bit more than i was expecting and i'm sure i screwed up an abstraction somewhere. advice on what to fix greatly appreciated.

otan committed:
schemachanger: drop schedule ID on DROP TABLE

This commit drops the schedule ID on the new schema changer during a
DROP TABLE.

Release note: None

ajwerner commented on pkg/sql/alter_table.go:
```diff
@@ -1740,6 +1493,412 @@ func (p *planner) updateFKBackReferenceName(
+	switch {
+	case before == nil && after == nil:
+		// Do not have to do anything here.
+	case before != nil && after != nil:
```
I think there's some thorny cases around removing the column but keeping TTL. Also, can we get some testing on multiple clauses where you remove the column via a reset and then add it back with a set in the same statement and the same transaction?

otan replied:
we will block allowing this behaviour

ajwerner reviewed (commented):
Can you add some testing around add a ttl expression that references a column added in the current transaction, added in the current statement. Then, to drive home my point, make adding that column fail by having an error in its default expression or check constraint. 

I think when we talk about the use of a mutation, we need to unpack the expected behavior in the face of other mutations and failure. The case where you currently don't use a mutation, which I think is troubling, is when you update an existing set of parameters. If another mutation in the same transaction fails, what should happen? I'm inclined to say that you actually do need a mutation and you need to make it richer.

ajwerner commented:
I propose that we add the limitation to these TTL changes that look a lot like the alter primary key changes whereby we don't allow any of these changes in a transaction and we don't allow them concurrent with any other schema changes. That will help us alleviate many (but not all) of my concerns.

otan commented on pkg/sql/schema_changer.go:
```diff
@@ -680,6 +701,10 @@ func (sc *SchemaChanger) exec(ctx context.Context) error {
 		return err
 	}
 
+	if err := sc.maybeUpdateScheduledJobsForRowLevelTTL(ctx, tableDesc); err != nil {
+		return err
+	}
```
this is accounting for DROP TABLE (one of the commits here). is there a better place for it? i can put it in `drop_table.go` if you prefer, but figured it should be as close to the DROP TABLE actually executing as possible.

otan commented on pkg/sql/alter_table.go:
```diff
@@ -1740,6 +1493,412 @@ func (p *planner) updateFKBackReferenceName(
 	return errors.Errorf("missing backreference for foreign key %s", ref.Name)
 }
 
+func dropColumnImpl(
```
yep, will happen

otan committed:
sql: prohibit dropping TTL automatic column

This commit prohibits dropping the TTL automatic column if TTL has been
defined on the table.

Release note: None

otan commented:
i've tacked on 3 commits at the end which blocks schema changes / ttl config changes whilst a ttl mutation is in progress

otan commented:
this is rebased & ready for review!

miretskiy commented on pkg/sql/descmetadata/metadata_updater.go:
```diff
@@ -145,3 +145,16 @@ func (mu metadataUpdater) SwapDescriptorSubComment(
+		"delete-schedule",
+		mu.txn,
+		sessiondata.InternalExecutorOverride{User: security.RootUserName()},
+		"DELETE FROM system.scheduled_jobs WHERE schedule_id = $1",
```
drop schedule perhaps?

otan replied:
i avoided using DROP SCHEDULE as it would allow users to DROP schedules as well. currently we've explicitly banned that. (also this was mostly form following the `deleteSchedule` code in the `sql` package; we have no hooks for this ... yet ...

ajwerner commented on pkg/sql/logictest/testdata/logic_test/row_level_ttl:
```diff
@@ -28,17 +28,145 @@ CREATE TABLE public.tbl (
+                                                                                           FAMILY fam_0_id_text_crdb_internal_expiration (id, text, crdb_internal_expiration)
+) WITH (ttl = 'on', ttl_automatic_column = 'on', ttl_expire_after = '10 days':::INTERVAL)
+
+statement error cannot modify TTL whilst another schema change is in progress
```
should we make this error clearer? It might be confusing because we're in an implicit transaction, also because it feels like it's all `ttl` related.

otan replied:
i've gone with `cannot modify TTL settings while another schema change on the table is being processed`

ajwerner commented on pkg/sql/schemachanger/scplan/internal/opgen/opgen_relation_depended_on_by.go:
```diff
@@ -31,7 +31,8 @@ func init() {
 				revertible(false),
 				emit(func(this *scpb.RelationDependedOnBy) scop.Op {
-					return &scop.RemoveRelationDependedOnBy{
+					return &scop.
+						RemoveRelationDependedOnBy{
```
?

ajwerner commented on pkg/sql/schemachanger/scdeps/sctestdeps/test_deps.go:
```diff
@@ -906,6 +906,12 @@ func (s *TestState) SwapDescriptorSubComment(
 
+// DeleteSchedule implements scexec.DescriptorMetadataUpdater
+func (s *TestState) DeleteSchedule(ctx context.Context, id int64) error {
+	s.LogSideEffectf("delete schedule ID %d", id)
```
now that you've added this, can we get a nice data-driven test in `schemachanger/testdata`. You do, `setup` and statements to set up the world and then `test` for the drop of the table and let it rewrite.

otan replied:
done. couldn't print out the schedule ID here though as it is non deterministic.

ajwerner reviewed (approved):
👍 LGTM mod minor things

otan committed:
sql: block schema changes whilst adding or dropping TTL

This commit blocks other schema changes whilst TTL is running, analagous
to ALTER COLUMN TYPE / ALTER PRIMARY KEY. This saves the state space to
think about in conjunction with other schema changes.

We also block changes to TTL whilst another schema change is running for
similar reasons.

Release note: None

otan reviewed (commented):
thanks for the review!

rafiss commented on pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_drop_column.go:
```diff
@@ -50,6 +52,14 @@ func alterTableDropColumn(b BuildCtx, table catalog.TableDescriptor, t *tree.Alt
 	}
 
+	if t.Column == colinfo.TTLDefaultExpirationColumnName && table.HasRowLevelTTL() {
+		panic(pgerror.Newf(
```
maybe `errors.WithHint(..., "use RESET(ttl)")`?

rafiss commented on pkg/sql/alter_table.go:
```diff
@@ -469,6 +469,14 @@ func (n *alterTableNode) startExec(params runParams) error {
 			}
 
+			if t.Column == colinfo.TTLDefaultExpirationColumnName && n.tableDesc.HasRowLevelTTL() {
+				return pgerror.Newf(
```
maybe `errors.WithHint(..., "use RESET(ttl)")`?

rafiss reviewed (approved):
reviewed what i can, but deferring to sql-schema on all the schema change logic.

one thought: do we need to make sure the TTL column is not referenced in a computed col or partial index def?

otan commented:
thanks!

> one thought: do we need to make sure the TTL column is not referenced in a computed col or partial index def?

answered in the other PR!

otan commented:
bors r=rafiss,ajwerner

need to run, tests are looking green, sorry if i interrupt your bors run

otan commented:
i love bors silently failing due to a rebase conflict

bors r=rafiss,ajwerner
